### PR TITLE
Issue 82 automated testing

### PR DESCRIPTION
**Pull request by JustinKyleJames**
This is a first pass at automated testing for the Globus plugin.

Some notes:

1. Right now this only works against an Ubuntu20 installation of Globus.  The test cases should run fine against others as long as the servers are set up properly.  I have docker scripts that will set up other OS's so for now I will run the tests manually within those containers.

2. At the moment I have to use a lot of telnet and ftp in the tests.  Some of it is intended and some is a byproduct of its use elsewhere.  This is due to the following:
    - Some things can't be done in globus-url-copy.
    - I am not sure how to set it up the certificates so that globus-url-copy can be run against the non-privileged user user1 without modifying a configuration file and restarting the server, so all tests against that user must use telnet/ftp.
    - I only know how to use telnet/ftp with the anonymous ftp user which is defined in the configuration as user1.
    - Right now the testing framework doesn't allow me to run icommands as other users (as it allows in the core irods tests).
    - Because of all of these, if I am testing something that requires telnet/ftp or I desire to test against a non-privileged user, I must use telnet/ftp to perform all actions even setup and cleanup actions.

I can solve these limitations in the future.

3. The tests/libs were simply copied from irods_client_s3_api with some S3 stuff removed.

### CODE REVIEW

File: README.md
Comment (trel) [inline README.md:338]: "runnign" -> "running"

File: tests/__init__.py
Comment (trel) [inline tests/__init__.py:1]: how about...

"For now, this test suite requires the iCommands packages."

File: tests/docker/build_globus.sh
Comment (trel) [inline tests/docker/build_globus.sh:2]: "irods-s3-api-builder" -> "irods-globus-connector-builder"

File: tests/docker/build_globus.sh
Comment (trel) [inline tests/docker/build_globus.sh:6]: "s3_api_source" -> "globus_connector_source"

File: tests/docker/build_globus.sh
Comment (trel) [inline tests/docker/build_globus.sh:8]: "irods-s3-api-builder" -> "irods-globus-connector-builder"

File: tests/docker/config.json
Comment (trel) [inline tests/docker/config.json:1]: this file doesn't seem correct at all...

File: tests/libs/six.py
Comment (trel) [inline tests/libs/six.py:1]: we need to include the entire six library ourselves?

**Review by trel — COMMENTED** [review]
very cool.   what's the status?  it all runs and passes and gives us confidence?

File: README.md
Comment (JustinKyleJames) [reply README.md:338]: done

File: tests/__init__.py
Comment (JustinKyleJames) [reply tests/__init__.py:1]: done

File: tests/docker/build_globus.sh
Comment (JustinKyleJames) [reply tests/docker/build_globus.sh:2]: This file was a copy from S3 API.  It is not needed or used.  I removed it.

File: tests/docker/build_globus.sh
Comment (JustinKyleJames) [reply tests/docker/build_globus.sh:6]: See above.

File: tests/docker/build_globus.sh
Comment (JustinKyleJames) [reply tests/docker/build_globus.sh:8]: See above.

File: tests/docker/config.json
Comment (JustinKyleJames) [reply tests/docker/config.json:1]: This file is not used or needed.  It has been removed.

**Review by alanking — COMMENTED** [review]
Just some preliminary observations... Haven't looked at the tests yet

File: tests/docker/build_globus.sh
Comment (JustinKyleJames) [reply tests/docker/build_globus.sh:1]: It has been removed.

File: tests/docker/docker-compose.yml
Comment (JustinKyleJames) [reply tests/docker/docker-compose.yml:9]: I believe so as the globus server has to connect to port 1247 on the irods server. ?

Comment (alanking) [reply tests/docker/docker-compose.yml:9]: Oh, this isn't exposing the port on the host machine. I was under the impression that any Docker containers connected via a Docker network could talk on any port unless you configure it otherwise. Maybe not?

File: tests/docker/globus.ubuntu20.Dockerfile
Comment (JustinKyleJames) [reply tests/docker/globus.ubuntu20.Dockerfile:1]: I renamed it.  The point was to test Ubuntu 20 for the Globus server but I'll have everything pointing to iRODS on Ubuntu 22.

Comment (JustinKyleJames) [reply tests/docker/globus.ubuntu20.Dockerfile:1]: I renamed the file.

File: tests/docker/irods.ubuntu20.Dockerfile
Comment (JustinKyleJames) [reply tests/docker/irods.ubuntu20.Dockerfile:1]: I updated the name of the Dockerfile.  We will have all the OS's for this one eventually.  (The iRODS one will remain Ubuntu 22.)

File: tests/docker/start.globus.ubuntu20.sh
Comment (JustinKyleJames) [reply tests/docker/start.globus.ubuntu20.sh:1]: Renamed to reflect that.

File: tests/docker/README.md
Comment (JustinKyleJames) [reply tests/docker/README.md:15]: This doesn't work for my version of docker so it might not for others.  I suggest leaving it as is.

Comment (alanking) [reply tests/docker/README.md:15]: Right-o

File: tests/libs/six.py
Comment (JustinKyleJames) [reply tests/libs/six.py:1]: I'm not sure.  If I try to install it with pip3 I get the following error.

```
bad magic number in 'libs.six': b'\x03\xf3\r\n'
```

Comment (trel) [reply tests/libs/six.py:1]: okay.   good enough for now - but we shouldn't need to have python2 compatibility soon (if not already).

Comment (JustinKyleJames) [reply tests/libs/six.py]: Now that I understand what six.py does and that we are only running this in Python 3, I was able to modify the libraries to remove any six.py dependencies.

**Comment by JustinKyleJames** [comment]
In addition to handling issues from above, I changed the checksum markers code to handle it if we got more than one marker.  This test now also handles the situation where we hang looking for a marker if the checksum comes back too fast.  In the latter case, we would still generate an error because we didn't see a marker but it wouldn't hang.

**Comment by JustinKyleJames** [comment]
> very cool. what's the status? it all runs and passes and gives us confidence?

Yes it runs and passes.  I believe it covers most of what I test.

Some things missing:

- Negative tests in general
- Testing persistent identifiers
- Testing gridmap_iRODS_callout.conf
- Verifying number of threads used for transfers based on configuration.  This requires parsing the log file.

**Comment by trel** [comment]
Consider making an issue for each known missing test (category).

File: tests/docker/docker-compose.yml
Comment (JustinKyleJames) [reply tests/docker/docker-compose.yml:9]: You might be right.  I will test it and remove if not needed.

Comment (JustinKyleJames) [reply tests/docker/docker-compose.yml]: The ports wasn't needed.  The tests pass without it.  I also logged onto the globus server and manually ran icommands.

This yml was based on the one for the S3 API.  I think for my own purposes I needed that port available externally,.

I removed the port section and pushed the commit.

**Comment by JustinKyleJames** [comment]
> Consider making an issue for each known missing test (category).

I created the issues.

File: tests/globus_test.py
Comment (alanking) [inline tests/globus_test.py:346]: Are these important/needed for this test? Should we make issues for these?

File: tests/globus_test.py
Comment (alanking) [inline tests/globus_test.py:337]: Would this hang be referring to the `while` loop below? Perhaps we could build in a timeout and a delay between checks, a la `delayAssert`? I suppose all that would do is introduce an intermittent failure rather than an intermittent hang... Is there anything we can do here to avoid the timing in the test, or is that just how it has to be (for now)?

File: tests/globus_test.py
Comment (alanking) [inline tests/globus_test.py:40]: Is there a way to check this in the test (e.g. stat some file, ask Globus, etc.)? Could be useful for understanding that particular failure should it arise.

File: tests/docker/README.md
Comment (alanking) [reply tests/docker/README.md:15]: Oh, or is the idea that you can leave the iRODS container running and run the Globus tests repeatedly on the same instance? That's pretty swell, if true.

File: tests/libs/__init__.py
Comment (JustinKyleJames) [reply tests/libs/__init__.py:1]: Yes, I will add that note.

File: tests/run_all_tests.py
Comment (JustinKyleJames) [reply tests/run_all_tests.py:19]: Right now it is just going to stdout.  I'll add an issue to write to a file.  Also see my second note below about the `docker-compose run globus` question.

File: tests/globus_test.py
Comment (JustinKyleJames) [reply tests/globus_test.py:337]: With the latest change it will not hang.  I will remove that comment.

File: tests/globus_test.py
Comment (JustinKyleJames) [reply tests/globus_test.py:40]: I can add that test.

File: tests/docker/README.md
Comment (JustinKyleJames) [reply tests/docker/README.md:15]: There are a couple of reasons I do this.

1. I just updated this to the following and the `docker-compose run X` allows me to target the specific instance.  So if I run `docker-compose run globus-ubuntu20` it will only start up the iRODS container (due to dependency) and the Ubuntu 20 Globus container.

```
    globus-ubuntu22:                                                                                                                                                                                       
        build:                                                                                                                                                                                             
            context: .                                                                                                                                                                                     
            dockerfile: globus.ubuntu22.Dockerfile                                                                                                                                                         
        depends_on:                                                                                                                                                                                        
            irods:                                                                                                                                                                                         
                condition: service_healthy                                                                                                                                                                 
        volumes:                                                                                                                                                                                           
            - ../../:/irods_client_globus_connector                                                                                                                                                        
    globus-ubuntu20:                                                                                                                                                                                       
        build:                                                                                                                                                                                             
            context: .                                                                                                                                                                                     
            dockerfile: globus.ubuntu20.Dockerfile                                                                                                                                                         
        depends_on:                                                                                                                                                                                        
            irods:                                                                                                                                                                                         
                condition: service_healthy                                                                                                                                                                 
        volumes:                                                                                                                                                                                           
            - ../../:/irods_client_globus_connector 
```
2. I'm not sure why but doing this doesn't interleave output from the two containers that are running (iRODS and Globus).  This makes it far easier to parse the results of the test.

File: tests/globus_test.py
Comment (JustinKyleJames) [reply tests/globus_test.py:346]: I don't think it is important necessarily.  That is more in line with how Globus does authentication via certificates.  It would make the tests cleaner though.  I'll add an issue on it but as far as I can tell it isn't a big priority.

File: tests/docker/README.md
Comment (JustinKyleJames) [reply tests/docker/README.md:15]: > Oh, or is the idea that you can leave the iRODS container running and run the Globus tests repeatedly on the same instance? That's pretty swell, if true.

It won't leave the container running.  If I want to leave it running I'll add a `tail -f /dev/null` just like if I used `docker-compose up`.  I'll probably accidentally check that in sooner or later.

File: tests/run_all_tests.py
Comment (JustinKyleJames) [reply tests/run_all_tests.py:19]: Added an issue on writing it to a file.

File: tests/libs/__init__.py
Comment (JustinKyleJames) [reply tests/libs/__init__.py:1]: Added a note and a README stating this.

File: tests/run_all_tests.py
Comment (trel) [reply tests/run_all_tests.py:19]: https://github.com/irods/irods_client_globus_connector/issues/96

File: tests/globus_test.py
Comment (JustinKyleJames) [reply tests/globus_test.py:346]: Added an issue on this.

File: tests/globus_test.py
Comment (JustinKyleJames) [reply tests/globus_test.py:337]: Updated the comment to state that the test will simply fail if we don't receive a marker.  I did verify this by using a small file.

File: tests/globus_test.py
Comment (trel) [reply tests/globus_test.py:346]: Please link it

File: tests/docker/README.md
Comment (alanking) [reply tests/docker/README.md:15]: Oh, I see. `up` has a similar option (specifying a service which will bring up all of its dependencies as well). I'm wondering whether we ought to separate the different platforms into their own compose files rather than having each platform defined as a service in the same file. I didn't catch that before.

Comment (JustinKyleJames) [reply tests/docker/README.md:15]: That was just in my latest commit.  I kind of like it the way it is rather than duplicating it.

Plus, the ability to not have stdout from the iRODS container is a big plus with `docker-compose run` vs `docker-compose up`.

File: tests/globus_test.py
Comment (JustinKyleJames) [reply tests/globus_test.py:40]: I added a test to check for error with that mapfile is not there.  I think this is more of a pure Globus test but it won't hurt.

Comment (alanking) [reply tests/globus_test.py:40]: Ah, I should have been more clear about my intention. The new test is fine, I guess, but I was thinking that such an assertion could be placed at the top of this test to ensure that the pre-requisite indicated by this comment is being met before attempting to run the test. That way, it is clear *why* the test failed (or could not be run, or whatever we decide to do with the fact that the resource mapfile wasn't set up properly) rather than running ahead in the test and failing with some possibly cryptic error.

I had something like this in mind, from the authentication tests which assume certain users with certain passwords exist before the test runs: https://github.com/irods/irods/blob/5e5057bdfff711163dcc6d5a26315997da1058a9/scripts/irods/test/test_auth.py#L33-L42

This might be overkill as a start, but just thought I would suggest it

**Comment by trel** [comment]
depends on https://github.com/irods/irods_client_globus_connector/pull/87 - merge that one first

**Comment by JustinKyleJames** [comment]
Note that this PR right now is dependent on #87 but only because there are a couple of test cases for the fix for #87.  However, I could reorganize the two to remove those two test cases and move them into the PR for #87.  Otherwise I will just rebase this after that one is merged.

**Comment by trel** [comment]
rebase seems fine - just keep the linking good and should be fine.

File: tests/globus_test.py
Comment (trel) [reply tests/globus_test.py:371]: https://github.com/irods/irods_client_globus_connector/issues/97

**Comment by korydraughn** [comment]
Is this ready?

**Comment by alanking** [comment]
#87 has been merged. Please rebase once more.

**Comment by JustinKyleJames** [comment]
> #87 has been merged. Please rebase once more.

Done

**Comment by korydraughn** [comment]
Missing issue number for commit `Adding Ubuntu20 tests`?

**Comment by JustinKyleJames** [comment]
> Missing issue number for commit `Adding Ubuntu20 tests`?

Yes, that along with many others were intended to be squashed.

**Review by korydraughn — APPROVED** [review]
Seems good to me.

Please wait for approval from @trel and @alanking.

**Review by trel — APPROVED** [review]

**Review by alanking — APPROVED** [review]
Let's # it.

**Comment by JustinKyleJames** [comment]
> Let's # it.

Done

**Pull request merged**
